### PR TITLE
ARROW-17832: [Python] Construct MapArray from sequence of dicts (instead of list of tuples)

### PR DESCRIPTION
**Pull request by kshitij12345**
Following snippet now works.
```python
dict_arr = pa.array([{'a': 1, 'b': 2}, {'c': 3}],
                    pa.map_(pa.string(), pa.int64()))
```

### CODE REVIEW

**Comment by kshitij12345** [comment]
cc: @AlenkaF @jorisvandenbossche

**Comment by kshitij12345** [comment]
MacOS failure looks irrelevant

Ref: https://github.com/apache/arrow/actions/runs/3348431666/jobs/5547482119#step:5:175

**Comment by kshitij12345** [comment]
Gentle ping @AlenkaF @jorisvandenbossche :)

**Comment by kshitij12345** [comment]
Gentle ping @AlenkaF @jorisvandenbossche :)

**Review by AlenkaF — APPROVED** [review]
LGTM. Thank you for all the contributions!

File: python/pyarrow/tests/test_array.py [hunk 1]
```diff
@@ -991,6 +991,18 @@ def test_map_labelled():
...
+def test_map_from_dict():
+    # ARROW-17832
+    tup_arr = pa.array([[('a', 1), ('b', 2)], [('c', 3)]],
+                       pa.map_(pa.string(), pa.int64()))
+    dict_arr = pa.array([{'a': 1, 'b': 2}, {'c': 3}],
+                        pa.map_(pa.string(), pa.int64()))
+
+    assert tup_arr.type.key_field == dict_arr.type.key_field
+    assert tup_arr.type.item_field == dict_arr.type.item_field
+    assert tup_arr == dict_arr
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_array.py:1003]:
```suggestion
    assert tup_arr.equals(dict_arr)
```

I think this should be sufficient for testing that it worked

File: python/pyarrow/src/arrow/python/python_to_arrow.cc [hunk 2]
```diff
@@ -760,6 +760,11 @@ class PyListConverter : public ListConverter<T, PyConverter, PyConverterTrait> {
       RETURN_NOT_OK(AppendSequence(value));
     } else if (PySet_Check(value) || (Py_TYPE(value) == &PyDictValues_Type)) {
       RETURN_NOT_OK(AppendIterable(value));
+    } else if (PyDict_Check(value) && this->options_.type->name() == "map") {
```
Comment (jorisvandenbossche) [inline python/pyarrow/src/arrow/python/python_to_arrow.cc:763]:
```suggestion
    } else if (PyDict_Check(value) && this->options_.type->id() == Type::MAP) {
```

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks, looks good!

**Comment by jorisvandenbossche** [comment]
Thanks!

File: python/pyarrow/tests/test_array.py [hunk 3]
```diff
@@ -991,6 +991,16 @@ def test_map_labelled():
     assert len(arr) == 2
 
 
+def test_map_from_dict():
+    # ARROW-17832
+    tup_arr = pa.array([[('a', 1), ('b', 2)], [('c', 3)]],
+                       pa.map_(pa.string(), pa.int64()))
+    dict_arr = pa.array([{'a': 1, 'b': 2}, {'c': 3}],
```
Comment (HyukjinKwon) [inline python/pyarrow/tests/test_array.py:998]: Thanks for this fix. Just had a quick followup question though. Should the nested dict work too? e.g.)

```python
import pyarrow as pa
pa.array([[{'a': 1, 'b': 2}, {'c': 3}]], pa.list_(pa.map_(pa.string(), pa.int64())))
```

It currently fails w/

```
pyarrow.lib.ArrowTypeError: Could not convert {'a': 1, 'b': 2} with type dict: was not a sequence or recognized null for conversion to list type
```

Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_array.py:998]: I would assume the nested case to work as well. Can you open a new issue about this?

Comment (HyukjinKwon) [reply python/pyarrow/tests/test_array.py:998]: Thanks guys. I filed an issue here: https://github.com/apache/arrow/issues/33928

**Comment by tommydangerous** [comment]
When will this fix be in PyPi?

**Comment by rok** [comment]
> When will this fix be in PyPi? Need the new version 12.0.0 to be installable using `pip` please pray

Releases are roughly every three months and 11.0.0 was released ~week ago.

**Comment by tommydangerous** [comment]
@rok how can we starting using this fix in Python immediately please?

**Comment by rok** [comment]
@tommydangerous until the release (I guess mid-April) you could [build pyarrow yourself](https://arrow.apache.org/docs/developers/python.html#building-on-linux-and-macos) or use [nightly builds](https://arrow.apache.org/docs/developers/python.html#installing-nightly-packages). Nightly builds are meant for testing only and not recommended for other uses.

**Comment by tommydangerous** [comment]
Thank you @rok , I’ll try the nightly builds.

**Pull request merged**